File: src/main/java/com/music/read/HomeView.java
package com.music.read;

import javafx.application.Platform;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.geometry.Insets;
import javafx.geometry.Orientation;
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.input.MouseButton;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.*;
import javafx.scene.paint.Color;
import javafx.stage.FileChooser;
import javafx.stage.Stage;
import javafx.util.Callback;

import java.io.File;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by xupanpan on 03/11/2017.
 */
public class HomeView {

    public ListView<MP3Info> listView;
    public ObservableList<MP3Info> list = FXCollections.observableArrayList();
    public boolean isSelectAll = true;
    private static final double WIDTH = 1000d;
    private static final double HIGTH = 700d;
    public Stage primaryStage;
    public VBox rootView;
    public static final String DEFULT_PRE = "-";
    public boolean nameFirst = true;
    public javafx.scene.control.TextField splitText;
    private javafx.scene.control.TextField deleteBeforText;
    private javafx.scene.control.TextField deleteAfterText;
    private javafx.scene.control.TextField text1;
    private javafx.scene.control.TextField text2;
    private BorderPane borderPane;

    private boolean isShowErrorCodeSettingView = false;
    private boolean isShowReplaceSettingView = false;
    private boolean isShowAddSettingView = false;
    private MenuBar menuBar;

    private MusicParser musicParser;
    private PlayManager playManager;
    private FileNameEditer fileNameEditer;

    public HomeView(Stage primaryStage) {
        this.primaryStage = primaryStage;
    }


    public void init() {
        musicParser = new MusicParser(this);
        playManager = new PlayManager(this);
        fileNameEditer = new FileNameEditer(this);
        primaryStage.setTitle("MP3信息助手");
        borderPane = new BorderPane();
        Scene scene = new Scene(borderPane, WIDTH, HIGTH);
        primaryStage.setScene(scene);
        primaryStage.show();

        initMenuBar();
        initView();
        borderPane.setCenter(rootView);
    }


    private void initMenuBar() {

        menuBar = new MenuBar();
        menuBar.setUseSystemMenuBar(true);
        Menu fileMenu = new Menu("文件");

        MenuItem openItem = new MenuItem("加载音乐");
        openItem.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent event) {
                FileChooser chooser = new FileChooser();
                chooser.getExtensionFilters().addAll(
                        new FileChooser.ExtensionFilter("All Images", "*.*"),
                        new FileChooser.ExtensionFilter("MP3", "*.mp3")
                );
                List<File> files = chooser.showOpenMultipleDialog(primaryStage);
                if (files != null) {
                    musicParser.loadMp3Data(files);
                }
            }
        });

        MenuItem clearItem = new MenuItem("删除选中");
        clearItem.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent event) {
                List<MP3Info> rList = new ArrayList<MP3Info>();
                for (int i = 0; i < list.size(); i++) {
                    MP3Info mp3Info = list.get(i);
                    if (mp3Info.isChecked) {
                        rList.add(mp3Info);
                    }
                }
                for (MP3Info info : rList) {
                    list.remove(info);
                }
            }
        });

        MenuItem clearAllItem = new MenuItem("清空列表");
        clearAllItem.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent event) {
                list.clear();
            }
        });

        MenuItem exitItem = new MenuItem("退出");
        exitItem.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent event) {
                Platform.exit();
            }
        });

        fileMenu.getItems().addAll(openItem, clearItem, clearAllItem, exitItem);


        Menu editMenu = new Menu("编辑");

        CheckMenuItem codeErrorItem = new CheckMenuItem("乱码分析");
        codeErrorItem.setSelected(isShowErrorCodeSettingView);
        codeErrorItem.selectedProperty().addListener(new ChangeListener<Boolean>() {
            public void changed(ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) {
                isShowErrorCodeSettingView = newValue;
                checkSettingView();
            }
        });


        CheckMenuItem replaceNameItem = new CheckMenuItem("批量替换名称字符");
        replaceNameItem.setSelected(isShowReplaceSettingView);
        replaceNameItem.selectedProperty().addListener(new ChangeListener<Boolean>() {
            public void changed(ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) {
                isShowReplaceSettingView = newValue;
                checkSettingView();
            }
        });

        CheckMenuItem addNameItem = new CheckMenuItem("批量添加名称字符");
        addNameItem.setSelected(isShowAddSettingView);
        addNameItem.selectedProperty().addListener(new ChangeListener<Boolean>() {
            public void changed(ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) {
                isShowAddSettingView = newValue;
                checkSettingView();

            }
        });


        editMenu.getItems().addAll(codeErrorItem, replaceNameItem, addNameItem);


        Menu helpMenu = new Menu("帮助");
        MenuItem useInfoItem = new MenuItem("使用说明");
        useInfoItem.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent event) {
            }
        });
        MenuItem aboutItem = new MenuItem("关于");
        aboutItem.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent event) {
            }
        });

        helpMenu.getItems().addAll(useInfoItem, aboutItem);

        menuBar.getMenus().addAll(fileMenu, editMenu, helpMenu);

        initPlayerView();

        borderPane.setTop(menuBar);
    }


    private void initPlayerView() {

        Menu playMenu = new Menu("play");
        menuBar.getMenus().addAll(playMenu);

    }


    private HBox getErrorCodeSetting() {

        HBox mHBox = new HBox(5);
        mHBox.setAlignment(Pos.CENTER_LEFT);
        Label splitLabel = new Label("文件分隔线:");

        splitText = new TextField(DEFULT_PRE);
        splitText.setAlignment(Pos.CENTER);
        splitText.setPrefWidth(40);


        CheckBox headBox = new CheckBox("歌手名在前");
        headBox.setSelected(nameFirst);
        headBox.selectedProperty().addListener(new ChangeListener<Boolean>() {
            public void changed(ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) {
                nameFirst = newValue;
            }
        });


        Button btn_ana = new Button("开始分析");
        btn_ana.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent event) {
                musicParser.executList();
            }
        });

        mHBox.getChildren().addAll(splitLabel, splitText,
                headBox, btn_ana);

        return mHBox;
    }


    private HBox getReplaceSetting() {

        HBox mHBox = new HBox(5);
        mHBox.setAlignment(Pos.CENTER_LEFT);
        Label petchDeletLabel = new Label("批量改文件名:");

        deleteBeforText = new TextField();
        deleteBeforText.setPromptText("改前");
        deleteBeforText.setPrefWidth(50);

        deleteAfterText = new TextField();
        deleteAfterText.setPromptText("改后");
        deleteAfterText.setPrefWidth(50);

        Button btn_delete = new Button("开始更改");
        btn_delete.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent event) {

                String textBefor = deleteBeforText.getText();
                String textAfter = deleteAfterText.getText();

                fileNameEditer.renameList(textBefor,textAfter);
            }
        });


        mHBox.getChildren().addAll(petchDeletLabel, deleteBeforText,
                deleteAfterText, btn_delete);

        return mHBox;
    }


    private HBox getAddSetting() {
        HBox mHBox = new HBox(5);
        mHBox.setAlignment(Pos.CENTER_LEFT);
        Label label1 = new Label("在第");
        text1 = new TextField();
        text1.setPrefWidth(30);
        Label label2 = new Label("位前，添加");
        text2 = new TextField();
        text2.setPrefWidth(50);

        Button btn_add = new Button("开始添加");
        btn_add.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent event) {
                String textWei = text1.getText();
                String textStr = text2.getText();
                fileNameEditer.addStrList(textWei,textStr);
            }
        });

        mHBox.getChildren().addAll(label1, text1,
                label2, text2, btn_add);

        return mHBox;
    }

    private Separator getVSeparator() {
        return new Separator(Orientation.VERTICAL);
    }

    private Separator getHSeparator() {
        return new Separator(Orientation.HORIZONTAL);
    }


    private HBox errorCodeSetting, replaceSetting, addSetting;
    private Separator vSeparator, vSeparator1, hSeparator;
    private HBox headSettingView;

    private void initSettingView() {
        headSettingView = new HBox(5);
        headSettingView.setAlignment(Pos.CENTER_LEFT);

        errorCodeSetting = getErrorCodeSetting();
        replaceSetting = getReplaceSetting();
        addSetting = getAddSetting();

        vSeparator = getVSeparator();
        vSeparator1 = getVSeparator();
        hSeparator = getHSeparator();

        headSettingView.getChildren().addAll(errorCodeSetting, vSeparator, replaceSetting, vSeparator1, addSetting);
        rootView.getChildren().addAll(headSettingView, hSeparator);
        checkSettingView();
    }

    private void checkSettingView() {

        errorCodeSetting.setVisible(isShowErrorCodeSettingView);
        errorCodeSetting.setManaged(isShowErrorCodeSettingView);

        replaceSetting.setVisible(isShowReplaceSettingView);
        replaceSetting.setManaged(isShowReplaceSettingView);

        addSetting.setVisible(isShowAddSettingView);
        addSetting.setManaged(isShowAddSettingView);

        if ((isShowErrorCodeSettingView && isShowReplaceSettingView) || (isShowErrorCodeSettingView && isShowAddSettingView)) {
            vSeparator.setVisible(true);
            vSeparator.setManaged(true);
        } else {
            vSeparator.setVisible(false);
            vSeparator.setManaged(false);
        }

        if (isShowReplaceSettingView && isShowAddSettingView) {
            vSeparator1.setVisible(true);
            vSeparator1.setManaged(true);
        } else {
            vSeparator1.setVisible(false);
            vSeparator1.setManaged(false);
        }

        if (isShowErrorCodeSettingView || isShowReplaceSettingView || isShowAddSettingView) {
            headSettingView.setVisible(true);
            headSettingView.setManaged(true);
            hSeparator.setVisible(true);
            hSeparator.setManaged(true);
        } else {
            headSettingView.setVisible(false);
            headSettingView.setManaged(false);
            hSeparator.setVisible(false);
            hSeparator.setManaged(false);
        }

    }


    private void initView() {

        rootView = new VBox();
        rootView.setPadding(new Insets(10, 10, 10, 10));
        rootView.setSpacing(10);
        rootView.setAlignment(Pos.TOP_CENTER);
        rootView.setBackground(Background.EMPTY);

        initSettingView();

        listView = new ListView<MP3Info>();
        listView.setItems(list);
        listView.setEditable(true);
        listView.setOrientation(Orientation.VERTICAL);
        listView.setCellFactory(new Callback<ListView<MP3Info>, ListCell<MP3Info>>() {
            public ListCell<MP3Info> call(ListView<MP3Info> param) {
                return new MP3ListCell();
            }
        });
        listView.getSelectionModel().selectedItemProperty().addListener(new ChangeListener<MP3Info>() {
            public void changed(ObservableValue<? extends MP3Info> observable, MP3Info oldValue, MP3Info newValue) {

            }
        });
        listView.setOnMouseClicked(new EventHandler<MouseEvent>() {
            public void handle(MouseEvent event) {
                if (event.getButton() == MouseButton.PRIMARY) {
                    if (event.getClickCount() == 2) {
                        MP3Info selectedItem = listView.getSelectionModel().getSelectedItem();
                        playManager.play(selectedItem);
                    }
                }
            }
        });
        VBox.setVgrow(listView, Priority.ALWAYS);
        rootView.getChildren().addAll(getHeadTitle(), listView);
    }

    private HBox getHeadTitle() {
        HBox itemView = new HBox(10);
        itemView.setAlignment(Pos.CENTER_LEFT);
        itemView.setPadding(new Insets(0, 9, 0, 9));
        CheckBox cb = new CheckBox();
        cb.setSelected(isSelectAll);
        cb.selectedProperty().addListener(new ChangeListener<Boolean>() {
            public void changed(ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) {
                isSelectAll = newValue;
                if (!list.isEmpty()) {
                    for (MP3Info mp3Info : list) {
                        mp3Info.isChecked = newValue;
                    }
                    listView.setItems(null);
                    listView.setItems(list);
                }
            }
        });

        Label name = new Label("文件名称");
        name.setPrefWidth(200);
        Label index = new Label("#");
        index.setPrefWidth(25);

        Label title = new Label("标题");
        title.setPrefWidth(140);
        Label artist = new Label("歌手");
        artist.setPrefWidth(80);
        Label album = new Label("专辑名称");
        album.setPrefWidth(160);
        Label time = new Label("时长");
        time.setPrefWidth(40);

        Separator s2 = new Separator(Orientation.VERTICAL);
        Separator s3 = new Separator(Orientation.VERTICAL);
        Separator s4 = new Separator(Orientation.VERTICAL);
        Separator s5 = new Separator(Orientation.VERTICAL);
        itemView.getChildren().addAll(cb, index, name, s2, title, s3, artist, s4, album, s5, time);
        initPlayControlView(itemView);
        return itemView;
    }

    public Label playTimeLabel;
    public Button playStop;

    private void initPlayControlView(HBox itemView) {
        Separator s6 = new Separator(Orientation.VERTICAL);
        playTimeLabel = new Label();
        playTimeLabel.setPrefWidth(45);
        playTimeLabel.setTextFill(Color.LIGHTSKYBLUE);

        playStop = new Button();
        setButtonPlay();
        playStop.setBackground(null);
        playStop.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent event) {
                onPlayClick();
            }
        });

        itemView.getChildren().addAll(s6, playStop, playTimeLabel);
    }


    public void setButtonPlay() {
        ImageView startImage = getStartImage();
        if (startImage != null) {
            playStop.setGraphic(startImage);
        }
    }
    public void setButtonStop() {
        ImageView stopImage = getStopImage();
        if (stopImage != null) {
            playStop.setGraphic(stopImage);
        }
    }

    private ImageView imageView = null;

    private ImageView getStartImage() {
        if (imageView == null) {
            try {
                InputStream resourceAsStream = HomeView.class.getClass().getResourceAsStream("/res/ic_live_play.png");
                Image im = new Image(resourceAsStream);
                imageView = new ImageView(im);
            } catch (Exception e) {
            }
        }
        return imageView;
    }

    private ImageView imageView1 = null;

    private ImageView getStopImage() {
        if (imageView1 == null) {
            try {
                InputStream resourceAsStream = HomeView.class.getClass().getResourceAsStream("/res/ic_live_suspend.png");
                Image im = new Image(resourceAsStream);
                imageView1 = new ImageView(im);
            } catch (Exception e) {
            }
        }
        return imageView1;
    }

    private void onPlayClick() {

        if (list.isEmpty()) return;

        if (playManager.currentPosition == -1) {
            playManager.play(0);
        } else {
            playManager.stopAndStart();
        }
    }



}

File: src/com/music/read/HomeView.kt
package com.music.read

import com.jfoenix.controls.JFXButton
import com.jfoenix.controls.JFXCheckBox
import com.jfoenix.controls.JFXListView
import com.jfoenix.controls.JFXTextField
import com.sun.javafx.scene.control.skin.ListViewSkin
import com.sun.javafx.scene.control.skin.VirtualFlow
import com.sun.org.apache.bcel.internal.generic.DADD
import javafx.application.Platform
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.event.EventType
import javafx.geometry.Insets
import javafx.geometry.Orientation
import javafx.geometry.Pos
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.image.Image
import javafx.scene.image.ImageView
import javafx.scene.input.*
import javafx.scene.layout.*
import javafx.scene.paint.Color
import javafx.stage.FileChooser
import javafx.stage.Stage
import javafx.stage.WindowEvent
import javafx.util.Callback

import java.io.File
import java.io.InputStream
import java.util.Timer
import java.util.TimerTask

/**
 * Created by xupanpan on 03/11/2017.
 */
class HomeView(var primaryStage: Stage) {

    var listView: JFXListView<MP3Info>

    var isSelectAll = true
    var rootView: VBox
    var nameFirst = true
    var splitText: javafx.scene.control.TextField
    private var deleteBeforText: javafx.scene.control.TextField? = null
    private var deleteAfterText: javafx.scene.control.TextField? = null
    private var text1: javafx.scene.control.TextField? = null
    private var text2: javafx.scene.control.TextField? = null
    private var borderPane: BorderPane? = null

    private var isShowErrorCodeSettingView = false
    private var isShowReplaceSettingView = false
    private var isShowAddSettingView = false
    private var menuBar: MenuBar? = null

    var musicParser: MusicFileParser
    var playManager: PlayManager
    var fileNameEditer: FileNameEditer
    private var playConfig: Config? = null
    private var image: Image? = null


    private var errorCodeSetting: HBox? = null
    private var replaceSetting: HBox? = null
    private var addSetting: HBox? = null
    private var vSeparator: Separator? = null
    private var vSeparator1: Separator? = null
    private var hSeparator: Separator? = null
    private var headSettingView: HBox? = null


    private var first = 0
    private var last = 0


    private val headTitle: HBox
        get() {
            val itemView = HBox(10.0)
            itemView.alignment = Pos.CENTER_LEFT
            itemView.padding = Insets(5.0, 9.0, 5.0, 9.0)
            val cb = JFXCheckBox()
            cb.isSelected = isSelectAll
            cb.selectedProperty().addListener { observable, oldValue, newValue ->
                isSelectAll = newValue!!
                DataManager.instans.setAllCheckStatus(newValue)
                listView.setItems(null)
                listView.setItems(DataManager.instans.list)
            }

            val name = Label("文件名称")
            name.prefWidth = 200.0
            val index = Label("#")
            index.prefWidth = 25.0

            val title = Label("标题")
            title.prefWidth = 140.0
            val artist = Label("歌手")
            artist.prefWidth = 80.0
            val album = Label("专辑名称")
            album.prefWidth = 160.0
            val time = Label("时长")
            time.prefWidth = 40.0

            val s2 = Separator(Orientation.VERTICAL)
            val s3 = Separator(Orientation.VERTICAL)
            val s4 = Separator(Orientation.VERTICAL)
            val s5 = Separator(Orientation.VERTICAL)
            itemView.children.addAll(cb, index, name, s2, title, s3, artist, s4, album, s5, time)
            initPlayControlView(itemView)
            return itemView
        }

    var playTimeLabel: Label
    var playStop: Button

    private var imageView: ImageView? = null

    private val startImage: ImageView?
        get() {
            if (imageView == null) {
                try {
                    val resourceAsStream = HomeView::class.java!!.javaClass.getResourceAsStream("/res/ic_live_play.png")
                    val im = Image(resourceAsStream)
                    imageView = ImageView(im)
                } catch (e: Exception) {
                }

            }
            return imageView
        }

    private var imageView1: ImageView? = null

    private val stopImage: ImageView?
        get() {
            if (imageView1 == null) {
                try {
                    val resourceAsStream = HomeView::class.java!!.javaClass.getResourceAsStream("/res/ic_live_suspend.png")
                    val im = Image(resourceAsStream)
                    imageView1 = ImageView(im)
                } catch (e: Exception) {
                }

            }

            return imageView1
        }

    init {

        try {
            image = Image(this.javaClass.getResourceAsStream("/res/icon.png"))
            primaryStage.icons.add(image)
        } catch (e: Exception) {
        }

        primaryStage.onCloseRequest = EventHandler { exitApp() }

    }


    fun exitApp() {

        val config = Config()
        config.isCheckedAll = isSelectAll
        config.playType = playManager.playType

        PlayListManager.savePlayConfig(config)
        PlayListManager.savePlayList(DataManager.instans.list)
        System.exit(0)
    }


    fun init() {
        musicParser = MusicFileParser(this)
        playManager = PlayManager(this)
        fileNameEditer = FileNameEditer(this)
        setCurrentPlayTitle(Main.APP_NAME)

        playConfig = PlayListManager.playConfig
        if (playConfig != null) {
            isSelectAll = playConfig!!.isCheckedAll
            playManager.playType = playConfig!!.playType
        }

        borderPane = BorderPane()
        val scene = Scene(borderPane!!, WIDTH, HIGTH)
        primaryStage.scene = scene
        primaryStage.show()

        initMenuBar()
        initView()
        borderPane!!.center = rootView

        initData()
    }

    private fun initData() {
        Thread(Runnable {
            val historyPlayList = PlayListManager.historyPlayList
            if (historyPlayList != null && !historyPlayList!!.isEmpty()) {
                DataManager.instans.add2List(historyPlayList)
                Timer().schedule(object : TimerTask() {
                    override fun run() {
                        Platform.runLater { playManager.stopAndStart() }
                    }
                }, 3000)
            }
        }).start()
    }

    fun setCurrentPlayTitle(title: String) {
        primaryStage.title = title
    }

    private fun initMenuBar() {

        menuBar = MenuBar()
        menuBar!!.isUseSystemMenuBar = true
        val fileMenu = Menu("文件")

        val openItem = MenuItem("加载音乐")
        openItem.onAction = EventHandler {
            val chooser = FileChooser()
            chooser.extensionFilters.addAll(
                    FileChooser.ExtensionFilter("All Images", "*.*")
            )
            val files = chooser.showOpenMultipleDialog(primaryStage)
            if (files != null) {
                musicParser.loadMp3Data(files)
            }
        }

        val clearItem = MenuItem("删除选中")
        clearItem.onAction = EventHandler {
            if (DataManager.instans.removeSelected()) {
                setCurrentPlayTitle(Main.APP_NAME)
                playManager.closePlay()
            }
        }

        val clearAllItem = MenuItem("清空列表")
        clearAllItem.onAction = EventHandler {
            DataManager.instans.clearList()
            setCurrentPlayTitle(Main.APP_NAME)
            playManager.closePlay()
            listView.userData = null
        }

        val exitItem = MenuItem("退出")
        exitItem.onAction = EventHandler { exitApp() }

        fileMenu.items.addAll(openItem, clearItem, clearAllItem, exitItem)


        val editMenu = Menu("编辑")

        val codeErrorItem = CheckMenuItem("乱码分析")
        codeErrorItem.isSelected = isShowErrorCodeSettingView
        codeErrorItem.selectedProperty().addListener { observable, oldValue, newValue ->
            isShowErrorCodeSettingView = newValue!!
            checkSettingView()
        }


        val replaceNameItem = CheckMenuItem("批量替换名称字符")
        replaceNameItem.isSelected = isShowReplaceSettingView
        replaceNameItem.selectedProperty().addListener { observable, oldValue, newValue ->
            isShowReplaceSettingView = newValue!!
            checkSettingView()
        }

        val addNameItem = CheckMenuItem("批量添加名称字符")
        addNameItem.isSelected = isShowAddSettingView
        addNameItem.selectedProperty().addListener { observable, oldValue, newValue ->
            isShowAddSettingView = newValue!!
            checkSettingView()
        }


        editMenu.items.addAll(codeErrorItem, replaceNameItem, addNameItem)


        val helpMenu = Menu("帮助")
        val useInfoItem = MenuItem("使用说明")
        useInfoItem.onAction = EventHandler { }
        val aboutItem = MenuItem("关于")
        aboutItem.onAction = EventHandler {
            val alert = Alert(Alert.AlertType.INFORMATION)
            alert.headerText = "关于" + Main.APP_NAME
            alert.contentText = "当前版本：" + Main.APP_VERSION
            alert.show()
        }

        helpMenu.items.addAll(useInfoItem, aboutItem)

        menuBar!!.menus.addAll(fileMenu, editMenu, helpMenu)

        initPlayerView()

        borderPane!!.top = menuBar
    }


    private fun initPlayerView() {

        val playMenu = Menu("play")

        val lastItem = MenuItem("上一曲")
        lastItem.onAction = EventHandler { playManager.playLast() }

        val playCloseItem = MenuItem("播放/停止")
        playCloseItem.onAction = EventHandler { onPlayClick() }

        val nextItem = MenuItem("下一曲")
        nextItem.onAction = EventHandler { playManager.playNext() }

        val toggleGroup = ToggleGroup()

        val singleItem = RadioMenuItem("单曲循环")
        singleItem.onAction = EventHandler { playManager.playType = PlayManager.PlayType.SINGLE }

        singleItem.toggleGroup = toggleGroup

        val shunItem = RadioMenuItem("顺序播放")
        shunItem.onAction = EventHandler { playManager.playType = PlayManager.PlayType.RECYCLE }
        shunItem.toggleGroup = toggleGroup

        val suijiItem = RadioMenuItem("随机播放")
        suijiItem.onAction = EventHandler { playManager.playType = PlayManager.PlayType.RADOM }
        suijiItem.toggleGroup = toggleGroup

        if (playConfig != null) {
            when (playConfig!!.playType) {
                PlayManager.PlayType.SINGLE -> singleItem.isSelected = true
                PlayManager.PlayType.RECYCLE -> shunItem.isSelected = true
                PlayManager.PlayType.RADOM -> suijiItem.isSelected = true
            }
        }

        playMenu.items.addAll(lastItem, playCloseItem, nextItem, SeparatorMenuItem(), singleItem, shunItem, suijiItem)
        menuBar!!.menus.addAll(playMenu)

    }


    private fun getErrorCodeSetting(): HBox {

        val mHBox = HBox(5.0)
        mHBox.alignment = Pos.CENTER_LEFT
        val splitLabel = Label("文件分隔线:")

        splitText = JFXTextField(DEFULT_PRE)
        splitText.alignment = Pos.CENTER
        splitText.prefWidth = 40.0


        val headBox = JFXCheckBox("歌手名在前")
        headBox.isSelected = nameFirst
        headBox.selectedProperty().addListener { observable, oldValue, newValue -> nameFirst = newValue!! }


        val btn_ana = JFXButton("开始分析")
        btn_ana.buttonType = JFXButton.ButtonType.RAISED
        btn_ana.onAction = EventHandler { musicParser.executList() }

        mHBox.children.addAll(splitLabel, splitText,
                headBox, btn_ana)

        return mHBox
    }


    private fun getReplaceSetting(): HBox {

        val mHBox = HBox(5.0)
        mHBox.alignment = Pos.CENTER_LEFT
        val petchDeletLabel = Label("批量改文件名:")

        deleteBeforText = JFXTextField()
        deleteBeforText!!.promptText = "改前"
        deleteBeforText!!.prefWidth = 40.0

        deleteAfterText = JFXTextField()
        deleteAfterText!!.promptText = "改后"
        deleteAfterText!!.prefWidth = 40.0

        val btn_delete = JFXButton("开始更改")
        btn_delete.buttonType = JFXButton.ButtonType.RAISED
        btn_delete.onAction = EventHandler {
            val textBefor = deleteBeforText!!.text
            val textAfter = deleteAfterText!!.text

            fileNameEditer.renameList(textBefor, textAfter)
        }


        mHBox.children.addAll(petchDeletLabel, deleteBeforText,
                deleteAfterText, btn_delete)

        return mHBox
    }


    private fun getAddSetting(): HBox {
        val mHBox = HBox(5.0)
        mHBox.alignment = Pos.CENTER_LEFT
        val label1 = Label("在第")
        text1 = JFXTextField()
        text1!!.prefWidth = 40.0
        val label2 = Label("位前，添加")
        text2 = JFXTextField()
        text2!!.prefWidth = 40.0

        val btn_add = JFXButton("开始添加")
        btn_add.buttonType = JFXButton.ButtonType.RAISED
        btn_add.onAction = EventHandler {
            val textWei = text1!!.text
            val textStr = text2!!.text
            fileNameEditer.addStrList(textWei, textStr)
        }

        mHBox.children.addAll(label1, text1,
                label2, text2, btn_add)

        return mHBox
    }

    private fun getVSeparator(): Separator {
        return Separator(Orientation.VERTICAL)
    }

    private fun getHSeparator(): Separator {
        return Separator(Orientation.HORIZONTAL)
    }

    private fun initSettingView() {
        headSettingView = HBox()
        headSettingView!!.padding = Insets(8.0, 8.0, 8.0, 8.0)
        headSettingView!!.spacing = 8.0
        headSettingView!!.alignment = Pos.CENTER_LEFT

        errorCodeSetting = getErrorCodeSetting()
        replaceSetting = getReplaceSetting()
        addSetting = getAddSetting()

        vSeparator = getVSeparator()
        vSeparator1 = getVSeparator()
        hSeparator = getHSeparator()

        headSettingView!!.children.addAll(errorCodeSetting, vSeparator, replaceSetting, vSeparator1, addSetting)
        rootView.children.addAll(headSettingView, hSeparator)
        checkSettingView()
    }

    private fun checkSettingView() {

        errorCodeSetting!!.isVisible = isShowErrorCodeSettingView
        errorCodeSetting!!.isManaged = isShowErrorCodeSettingView

        replaceSetting!!.isVisible = isShowReplaceSettingView
        replaceSetting!!.isManaged = isShowReplaceSettingView

        addSetting!!.isVisible = isShowAddSettingView
        addSetting!!.isManaged = isShowAddSettingView

        if (isShowErrorCodeSettingView && isShowReplaceSettingView || isShowErrorCodeSettingView && isShowAddSettingView) {
            vSeparator!!.isVisible = true
            vSeparator!!.isManaged = true
        } else {
            vSeparator!!.isVisible = false
            vSeparator!!.isManaged = false
        }

        if (isShowReplaceSettingView && isShowAddSettingView) {
            vSeparator1!!.isVisible = true
            vSeparator1!!.isManaged = true
        } else {
            vSeparator1!!.isVisible = false
            vSeparator1!!.isManaged = false
        }

        if (isShowErrorCodeSettingView || isShowReplaceSettingView || isShowAddSettingView) {
            headSettingView!!.isVisible = true
            headSettingView!!.isManaged = true
            hSeparator!!.isVisible = true
            hSeparator!!.isManaged = true
        } else {
            headSettingView!!.isVisible = false
            headSettingView!!.isManaged = false
            hSeparator!!.isVisible = false
            hSeparator!!.isManaged = false
        }

    }

    private fun initView() {

        rootView = VBox()
        rootView.alignment = Pos.TOP_CENTER
        rootView.background = Background.EMPTY

        initSettingView()

        listView = JFXListView()
        listView.setItems(DataManager.instans.list)
        listView.isEditable = true
        listView.border = FxViewUtil.getBorder(Color.WHITE, 0.0, 0.0)
        listView.background = FxViewUtil.getBackground(Color.WHITE, 0)
        listView.orientation = Orientation.VERTICAL
        listView.setCellFactory { MP3ListCell(this@HomeView) }
        listView.selectionModel.selectedItemProperty().addListener { observable, oldValue, newValue -> }

        listView.onMouseClicked = EventHandler { event ->
            if (event.button == MouseButton.PRIMARY) {
                if (event.clickCount == 2) {
                    val selectedItem = listView.selectionModel.selectedItem
                    playManager.play(selectedItem)
                }
            }
        }

        VBox.setVgrow(listView, Priority.ALWAYS)

        listView.onDragOver = EventHandler { event ->
            if (event.gestureSource !== listView) {
                event.acceptTransferModes(*TransferMode.ANY)
            }
        }

        listView.onDragDropped = EventHandler { event ->
            val dragboard = event.dragboard
            val files = dragboard.files
            if (files.size > 0) {
                musicParser.loadMp3Data(files)
            }
        }

        listView.onKeyReleased = EventHandler { event ->
            val code = event.code
            if (code == KeyCode.ENTER) {
                playManager.play(listView.selectionModel.selectedIndex)
            } else if (code == KeyCode.SPACE) {
                onPlayClick()
            } else if (code == KeyCode.RIGHT) {
                playManager.playNext()
            } else if (code == KeyCode.LEFT) {
                playManager.playLast()
            }
        }

        rootView.children.addAll(headTitle, listView)
    }

    private fun getFirstAndLast() {
        if (DataManager.instans.isListEmpty) {
            return
        }
        try {
            val ts = listView.skin as ListViewSkin<*>
            val vf = ts.children[0] as VirtualFlow<*>
            first = vf.firstVisibleCell.index
            last = vf.lastVisibleCell.index
        } catch (ex: Exception) {
            ex.printStackTrace()
        }

    }

    fun scrollToShow() {
        getFirstAndLast()
        val currentPlayPosition = DataManager.instans.currentPlayPosition
        if (currentPlayPosition < first || currentPlayPosition > last) {
            listView.scrollTo(currentPlayPosition)
        }
    }


    private fun showNotification() {

    }

    private fun initPlayControlView(itemView: HBox) {
        val s6 = Separator(Orientation.VERTICAL)
        playTimeLabel = Label()
        playTimeLabel.prefWidth = 45.0
        playTimeLabel.textFill = Color.LIGHTSKYBLUE
        playTimeLabel.onMouseClicked = EventHandler { event ->
            if (event.button == MouseButton.PRIMARY) {
                if (event.clickCount == 2) {

                    scrollToShow()
                }
            }
        }

        playStop = Button()
        setButtonPlay()
        playStop.background = null
        playStop.onAction = EventHandler { onPlayClick() }

        itemView.children.addAll(s6, playStop, playTimeLabel)
    }


    fun setButtonPlay() {
        val startImage = startImage
        if (startImage != null) {
            playStop.graphic = startImage
        }
    }

    fun setButtonStop() {
        val stopImage = stopImage
        if (stopImage != null) {
            playStop.graphic = stopImage
        }
    }

    fun onPlayClick() {

        if (DataManager.instans.isListEmpty) return

        if (DataManager.instans.currentPlayPosition == -1) {
            playManager.play(0)
        } else {
            playManager.stopAndStart()
        }
    }

    companion object {
        private val WIDTH = 1000.0
        private val HIGTH = 700.0
        val DEFULT_PRE = "-"
    }


}
